<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Verónica Lloréns Rico" />

<meta name="date" content="2018-09-11" />

<title>Simulating metatranscriptomes with metaester</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">Simulating metatranscriptomes with metaester</h1>
<h4 class="author"><em>Verónica Lloréns Rico</em></h4>
<h4 class="date"><em>2018-09-11</em></h4>



<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>Metaester is an R package to simulate metatranscriptomics datasets.</p>
<p>Metaester is built from <a href="https://github.com/alyssafrazee/polyester">Polyester</a>, an R package to simulate RNA-seq experiments, hence its name. Many of the functions in Metaester are wrappers to Polyester functions. However, we have not exploited all the possible functionalities of polyester yet.</p>
<p>Metaester allows to simulate an entire microbial community, by first distributing the RNA-seq reads among the different genomes present in the simulated sample and then distributing the RNA-seq reads among the different genes, within each simulated genome.</p>
</div>
<div id="installation" class="section level1">
<h1>Installation</h1>
<p>Metaester is built upon the devel version of Polyester (as of July 2018, not all changes in the devel version were pushed to the bioC version). Therefore, to use Metaester, we strongly recommend to install the devel version of Polyester:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;alyssafrazee/polyester&quot;</span>)</code></pre></div>
<p>To install Metaester, run the following code:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;vllorens/metaester&quot;</span>)</code></pre></div>
</div>
<div id="the-metaester-workflow" class="section level1">
<h1>The metaester workflow</h1>
<p>Metaester is built with the aim of simulating metatranscriptomics datasets as simply as possible. Only 4 steps are required to simulate an experiment:</p>
<ul>
<li>Step 0: Define the parameters of the experiment: number of reads, replicates, number of cases and controls for differential expression, etc.</li>
<li>Step 1: Generate genome read distribution: distribute the total reads <em>per sample</em> among the genomes to simulate</li>
<li>Step 2: For each genome, generate gene read distribution: distribute the total reads <em>per genome</em> among the genes of that genome</li>
<li>Step 3: Generate the fasta files with the raw reads of the simulation</li>
</ul>
</div>
<div id="metaester-examples" class="section level1">
<h1>Metaester examples</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(metaester)</code></pre></div>
<p>This first set of examples uses an empirical distribution for the community composition, modeled as a negative binomial distribution.</p>
<div id="example-1-10-replicates-no-differential-composition-no-differential-expression" class="section level3">
<h3>Example 1: 10 replicates, NO differential composition, NO differential expression</h3>
<div id="step-0-define-the-parameters-of-the-experiment" class="section level4">
<h4>Step 0: define the parameters of the experiment</h4>
<p>In this case, we will simulate 10 samples with the same microbial composition and gene expression profiles, so the only differences among them will be due to the process of sampling. Those would be analogous to technical replicates in sequencing.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load the genomes of 5 gut bacteria and list them</span>
path_to_fastas &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;metaester&quot;</span>, <span class="dt">mustWork =</span> <span class="ot">TRUE</span>) 
genomeList &lt;-<span class="st"> </span><span class="kw">list.files</span>(path_to_fastas, <span class="dt">pattern =</span> <span class="st">&quot;.genes.fna&quot;</span>)                 
genomeList &lt;-<span class="st"> </span><span class="kw">gsub</span>(genomeList, <span class="dt">pattern =</span> <span class="st">&quot;.genes.fna&quot;</span>, <span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>)             

genomeList
<span class="co">#&gt; [1] &quot;bobeum&quot;        &quot;efaecalis&quot;     &quot;fprausnitzii&quot;  &quot;ljohnsonii&quot;   </span>
<span class="co">#&gt; [5] &quot;rintestinalis&quot;</span>

<span class="co"># set random seed for the genome composition</span>
random_seed &lt;-<span class="st"> </span><span class="dv">1</span>      
                       
<span class="co"># set number of replicates and reads to simulate.</span>
<span class="co"># as there's no differential expression among samples no need to specify number of cases and controls</span>
samples &lt;-<span class="st"> </span><span class="dv">10</span>          
reads_perSample &lt;-<span class="st"> </span><span class="dv">10000</span>

<span class="co"># specify output directory</span>
output_simulation &lt;-<span class="st"> &quot;output_data&quot;</span></code></pre></div>
</div>
<div id="step-1-generate-genome-read-distribution." class="section level4">
<h4>Step 1: Generate genome read distribution.</h4>
<p>Here, we use the function <code>compositionGenomesMetaT()</code>. We will distribute the reads among the 5 genomes using an empirical distribution (i.e. a negative binomial, derived from real data).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genomeComp &lt;-<span class="st"> </span><span class="kw">compositionGenomesMetaT</span>(<span class="dt">composition =</span> <span class="st">&quot;empirical&quot;</span>, <span class="dt">empiricalSeed =</span> random_seed, 
                                      <span class="dt">nReplicates =</span> samples, <span class="dt">genomes =</span> genomeList, 
                                      <span class="dt">nReads =</span> reads_perSample)

<span class="co"># print the microbial composition table to see that differences in abundances are only due to sampling</span>
genomeComp
<span class="co">#&gt;               sample_1 sample_2 sample_3 sample_4 sample_5 sample_6</span>
<span class="co">#&gt; bobeum               1        2        1        0        4        4</span>
<span class="co">#&gt; efaecalis          123      124      126      140      138      128</span>
<span class="co">#&gt; fprausnitzii        27       33       28       27       22       30</span>
<span class="co">#&gt; ljohnsonii        8199     8227     8163     8112     8091     8145</span>
<span class="co">#&gt; rintestinalis     1650     1614     1682     1721     1745     1693</span>
<span class="co">#&gt;               sample_7 sample_8 sample_9 sample_10</span>
<span class="co">#&gt; bobeum               3        2        2         0</span>
<span class="co">#&gt; efaecalis          113      110      104       112</span>
<span class="co">#&gt; fprausnitzii        24       34       35        32</span>
<span class="co">#&gt; ljohnsonii        8216     8166     8176      8211</span>
<span class="co">#&gt; rintestinalis     1644     1688     1683      1645</span></code></pre></div>
</div>
<div id="step-2-generate-gene-read-distribution." class="section level4">
<h4>Step 2: Generate gene read distribution.</h4>
<p>Here, we use the function <code>simulateMetaTranscriptome()</code>. This function takes the reads assigned to each genome and the list of genes within this genome (extracted from the FASTA file), and then distributes the reads among the different genes of that genome. This function does this for the 5 genomes in our list and for the 10 samples and aggregates the output, resulting in a single gene count matrix, in which each row represents a single gene from any of this 5 bacteria and each column represents one of the 10 samples.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">metaT_experiment=<span class="kw">simulateMetaTranscriptome</span>(<span class="dt">genomeFileDir =</span> path_to_fastas, <span class="dt">genomeReadMatrix =</span> genomeComp)
<span class="co">#&gt; Generating data from baseline model.</span>
<span class="co">#&gt; Generating data from baseline model.</span>
<span class="co">#&gt; Generating data from baseline model.</span>
<span class="co">#&gt; Generating data from baseline model.</span>
<span class="co">#&gt; Generating data from baseline model.</span></code></pre></div>
</div>
<div id="step-3-generate-the-fasta-files." class="section level4">
<h4>Step 3: Generate the FASTA files.</h4>
<p>Here, we use the function <code>simulateFastaReads()</code>. This function takes the output from the previous one, as well as the gene FASTAS from each genome. It writes the simulated reads fasta files to the output directory (2 fasta files per sample in paired-end mode or 1 in single-end mode). It also writes the gene count matrix in the same directory, and a list of differentially expressed genes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Step 4: Generate the fasta files (function simulateFastaReads)
<span class="kw">simulateFastaReads</span>(<span class="dt">genomeFileDir =</span> path_to_fastas, <span class="dt">outdir =</span> output_simulation, 
                   <span class="dt">simulatedDataSet =</span> metaT_experiment)</code></pre></div>
<hr />
</div>
</div>
<div id="example-2-10-replicates-differential-composition-between-cases-and-controls-no-differential-expression" class="section level3">
<h3>Example 2: 10 replicates, differential composition (between cases and controls), NO differential expression</h3>
<div id="step-0-define-the-parameters-of-the-experiment-1" class="section level4">
<h4>Step 0: define the parameters of the experiment</h4>
<p>In this case, we will simulate 10 samples in two groups of 5, in which each group has a different microbial composition. Here, bacteria change their relative abundances from one group to another, but the functions they perform are the same, as there is no differential expression.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load the genomes of 5 gut bacteria and list them</span>
path_to_fastas &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;metaester&quot;</span>, <span class="dt">mustWork =</span> <span class="ot">TRUE</span>) 
genomeList &lt;-<span class="st"> </span><span class="kw">list.files</span>(path_to_fastas, <span class="dt">pattern =</span> <span class="st">&quot;.genes.fna&quot;</span>)                 
genomeList &lt;-<span class="st"> </span><span class="kw">gsub</span>(genomeList, <span class="dt">pattern =</span> <span class="st">&quot;.genes.fna&quot;</span>, <span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>)             

genomeList
<span class="co">#&gt; [1] &quot;bobeum&quot;        &quot;efaecalis&quot;     &quot;fprausnitzii&quot;  &quot;ljohnsonii&quot;   </span>
<span class="co">#&gt; [5] &quot;rintestinalis&quot;</span>

<span class="co"># set random seed for the genome composition (for 10 samples, we generate the 2 groups)</span>
random_seed &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>, <span class="dt">each=</span><span class="dv">5</span>)     
                       
<span class="co"># set number of replicates and reads to simulate.</span>
<span class="co"># as there's no differential expression among samples no need to specify number of cases and controls</span>
samples &lt;-<span class="st"> </span><span class="dv">10</span>          
reads_perSample &lt;-<span class="st"> </span><span class="dv">10000</span>

<span class="co"># specify output directory</span>
output_simulation &lt;-<span class="st"> &quot;output_data&quot;</span></code></pre></div>
</div>
<div id="step-1-generate-genome-read-distribution.-1" class="section level4">
<h4>Step 1: Generate genome read distribution.</h4>
<p>Here, we use the function <code>compositionGenomesMetaT()</code>. We will distribute the reads among the 5 genomes using an empirical distribution (i.e. a negative binomial, derived from real data).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genomeComp &lt;-<span class="st"> </span><span class="kw">compositionGenomesMetaT</span>(<span class="dt">composition =</span> <span class="st">&quot;empirical&quot;</span>, <span class="dt">empiricalSeed =</span> random_seed, 
                                      <span class="dt">nReplicates =</span> samples, <span class="dt">genomes =</span> genomeList, 
                                      <span class="dt">nReads =</span> reads_perSample)

<span class="co"># print the microbial composition table to see the differences between the two groups</span>
genomeComp
<span class="co">#&gt;               sample_1 sample_2 sample_3 sample_4 sample_5 sample_6</span>
<span class="co">#&gt; bobeum               1        2        1        0        4      436</span>
<span class="co">#&gt; efaecalis          123      124      126      140      138     1382</span>
<span class="co">#&gt; fprausnitzii        27       33       28       27       22       26</span>
<span class="co">#&gt; ljohnsonii        8199     8227     8163     8112     8091     5706</span>
<span class="co">#&gt; rintestinalis     1650     1614     1682     1721     1745     2450</span>
<span class="co">#&gt;               sample_7 sample_8 sample_9 sample_10</span>
<span class="co">#&gt; bobeum             388      401      407       384</span>
<span class="co">#&gt; efaecalis         1362     1387     1373      1369</span>
<span class="co">#&gt; fprausnitzii        22       28       30        25</span>
<span class="co">#&gt; ljohnsonii        5762     5705     5791      5758</span>
<span class="co">#&gt; rintestinalis     2466     2479     2399      2464</span></code></pre></div>
</div>
<div id="step-2-generate-gene-read-distribution.-1" class="section level4">
<h4>Step 2: Generate gene read distribution.</h4>
<p>Here, we use the function <code>simulateMetaTranscriptome()</code>. This function takes the reads assigned to each genome and the list of genes within this genome (extracted from the FASTA file), and then distributes the reads among the different genes of that genome. This function does this for the 5 genomes in our list, and for the 10 samples, and aggregates the output, resulting in a single gene count matrix, in which each row represents a single gene from any of this 5 bacteria and each column represents one of the 10 samples.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">metaT_experiment=<span class="kw">simulateMetaTranscriptome</span>(<span class="dt">genomeFileDir =</span> path_to_fastas, <span class="dt">genomeReadMatrix =</span> genomeComp)</code></pre></div>
</div>
<div id="step-3-generate-the-fasta-files.-1" class="section level4">
<h4>Step 3: Generate the FASTA files.</h4>
<p>Here, we use the function <code>simulateFastaReads()</code>. This function takes the output from the previous one, as well as the gene FASTAS from each genome. It writes the simulated reads fasta files to the output directory (2 fasta files per sample in paired-end mode or 1 in single-end mode). It also writes the gene count matrix in the same directory, and a list of differentially expressed genes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">simulateFastaReads</span>(<span class="dt">genomeFileDir =</span> path_to_fastas, <span class="dt">outdir =</span> output_simulation,
                   <span class="dt">simulatedDataSet =</span> metaT_experiment)</code></pre></div>
<hr />
</div>
</div>
<div id="example-3-10-replicates-differential-composition-across-all-samples-no-differential-expression" class="section level3">
<h3>Example 3: 10 replicates, differential composition across all samples, NO differential expression</h3>
<div id="step-0-define-the-parameters-of-the-experiment-2" class="section level4">
<h4>Step 0: define the parameters of the experiment</h4>
<p>In this case, we will simulate 10 samples, but each sample has a different bacterial composition. This would be equivalent to sequencing samples from different individuals. Here, bacteria change their relative abundances from one sample to another, but the functions they perform are the same, as there is no differential expression.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load the genomes of 5 gut bacteria and list them</span>
path_to_fastas &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;metaester&quot;</span>, <span class="dt">mustWork =</span> <span class="ot">TRUE</span>) 
genomeList &lt;-<span class="st"> </span><span class="kw">list.files</span>(path_to_fastas, <span class="dt">pattern =</span> <span class="st">&quot;.genes.fna&quot;</span>)                 
genomeList &lt;-<span class="st"> </span><span class="kw">gsub</span>(genomeList, <span class="dt">pattern =</span> <span class="st">&quot;.genes.fna&quot;</span>, <span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>)             

genomeList
<span class="co">#&gt; [1] &quot;bobeum&quot;        &quot;efaecalis&quot;     &quot;fprausnitzii&quot;  &quot;ljohnsonii&quot;   </span>
<span class="co">#&gt; [5] &quot;rintestinalis&quot;</span>

<span class="co"># set random seed for the genome composition (for 10 samples, a different seed per sample)</span>
random_seed &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="dv">10</span>     
                       
<span class="co"># set number of samples and reads to simulate.</span>
<span class="co"># as there's no differential expression among samples no need to specify number of cases and controls</span>
samples &lt;-<span class="st"> </span><span class="dv">10</span>          
reads_perSample &lt;-<span class="st"> </span><span class="dv">10000</span>

<span class="co"># specify output directory</span>
output_simulation &lt;-<span class="st"> &quot;output_data&quot;</span></code></pre></div>
</div>
<div id="step-1-generate-genome-read-distribution.-2" class="section level4">
<h4>Step 1: Generate genome read distribution.</h4>
<p>Here, we use the function <code>compositionGenomesMetaT()</code>. We will distribute the reads among the 5 genomes using an empirical distribution (i.e. a negative binomial, derived from real data).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genomeComp &lt;-<span class="st"> </span><span class="kw">compositionGenomesMetaT</span>(<span class="dt">composition =</span> <span class="st">&quot;empirical&quot;</span>, <span class="dt">empiricalSeed =</span> random_seed, 
                                      <span class="dt">nReplicates =</span> samples, <span class="dt">genomes =</span> genomeList, 
                                      <span class="dt">nReads =</span> reads_perSample)</code></pre></div>
</div>
<div id="step-2-generate-gene-read-distribution.-2" class="section level4">
<h4>Step 2: Generate gene read distribution.</h4>
<p>Here, we use the function <code>simulateMetaTranscriptome()</code>. This function takes the reads assigned to each genome and the list of genes within this genome (extracted from the FASTA file), and then distributes the reads among the different genes of that genome. This function does this for the 5 genomes in our list, and for the 10 samples, and aggregates the output, resulting in a single gene count matrix, in which each row represents a single gene from any of this 5 bacteria and each column represents one of the 10 samples.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">metaT_experiment=<span class="kw">simulateMetaTranscriptome</span>(<span class="dt">genomeFileDir =</span> path_to_fastas, <span class="dt">genomeReadMatrix =</span> genomeComp)</code></pre></div>
</div>
<div id="step-3-generate-the-fasta-files.-2" class="section level4">
<h4>Step 3: Generate the FASTA files.</h4>
<p>Here, we use the function <code>simulateFastaReads()</code>. This function takes the output from the previous one, as well as the gene FASTAS from each genome. It writes the simulated reads fasta files to the output directory (2 fasta files per sample in paired-end mode or 1 in single-end mode). It also writes the gene count matrix in the same directory, and a list of differentially expressed genes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">simulateFastaReads</span>(<span class="dt">genomeFileDir =</span> path_to_fastas, <span class="dt">outdir =</span> output_simulation,
                   <span class="dt">simulatedDataSet =</span> metaT_experiment)</code></pre></div>
<hr />
</div>
</div>
<div id="example-4-10-replicates-no-differential-composition-10-of-genes-are-differentially-expressed" class="section level3">
<h3>Example 4: 10 replicates, NO differential composition, <em>10% of genes are differentially expressed</em></h3>
<div id="step-0-define-the-parameters-of-the-experiment-3" class="section level4">
<h4>Step 0: define the parameters of the experiment</h4>
<p>In this case, we will simulate 10 samples with the same bacterial composition. However, we will divide them in between 5 cases and 5 controls. 10% of the genes of each bacterium will be differentially expressed between the two conditions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load the genomes of 5 gut bacteria and list them</span>
path_to_fastas &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;metaester&quot;</span>, <span class="dt">mustWork =</span> <span class="ot">TRUE</span>) 
genomeList &lt;-<span class="st"> </span><span class="kw">list.files</span>(path_to_fastas, <span class="dt">pattern =</span> <span class="st">&quot;.genes.fna&quot;</span>)                 
genomeList &lt;-<span class="st"> </span><span class="kw">gsub</span>(genomeList, <span class="dt">pattern =</span> <span class="st">&quot;.genes.fna&quot;</span>, <span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>)             

genomeList
<span class="co">#&gt; [1] &quot;bobeum&quot;        &quot;efaecalis&quot;     &quot;fprausnitzii&quot;  &quot;ljohnsonii&quot;   </span>
<span class="co">#&gt; [5] &quot;rintestinalis&quot;</span>

<span class="co"># set random seed for the genome composition (for 10 samples, a different seed per sample)</span>
random_seed &lt;-<span class="st"> </span><span class="dv">1</span>

<span class="co"># set number of samples and reads to simulate.</span>
<span class="co"># in this case there is differential expression, so we need to specify the number of cases and controls</span>
samples &lt;-<span class="st"> </span><span class="dv">10</span>         
nCases &lt;-<span class="st"> </span><span class="dv">5</span>
nControls &lt;-<span class="st"> </span><span class="dv">5</span>
reads_perSample &lt;-<span class="st"> </span><span class="dv">10000</span>

<span class="co"># differentially expresed genes will have any of these fold changes</span>
<span class="co"># X &gt; 1 overexpression</span>
<span class="co"># X &lt; 1 depletion</span>
<span class="co"># X = 1 no changes (it is important to include that some genes will have no changes)</span>
fold_changes=<span class="kw">c</span>(<span class="fl">0.25</span>,<span class="fl">0.5</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">4</span>) 

<span class="co"># after specifying fold changes, it is required to specify the probability of each of them in order</span>
<span class="co"># probabilities don't need to add up to 100, but it makes it easier to interpret. In this case: </span>
<span class="co"># 2.5% probability of 0.25-fold</span>
<span class="co"># 2.5% probability of 0.5-fold</span>
<span class="co"># 90% probability of 1-fold (no changes)</span>
<span class="co"># 2.5% probability of 2-fold</span>
<span class="co"># 2.5% probability of 4-fold</span>
fold_probs=<span class="kw">c</span>(<span class="fl">2.5</span>,<span class="fl">2.5</span>,<span class="dv">90</span>,<span class="fl">2.5</span>,<span class="fl">2.5</span>) 

output_simulation &lt;-<span class="st"> &quot;output_data&quot;</span></code></pre></div>
</div>
<div id="step-1-generate-genome-read-distribution.-3" class="section level4">
<h4>Step 1: Generate genome read distribution.</h4>
<p>Here, we use the function <code>compositionGenomesMetaT()</code>. We will distribute the reads among the 5 genomes using an empirical distribution (i.e. a negative binomial, derived from real data).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genomeComp &lt;-<span class="st"> </span><span class="kw">compositionGenomesMetaT</span>(<span class="dt">composition =</span> <span class="st">&quot;empirical&quot;</span>, <span class="dt">empiricalSeed =</span> random_seed, 
                                      <span class="dt">nReplicates =</span> samples, <span class="dt">genomes =</span> genomeList, 
                                      <span class="dt">nReads =</span> reads_perSample)</code></pre></div>
</div>
<div id="step-2-generate-gene-read-distribution.-3" class="section level4">
<h4>Step 2: Generate gene read distribution.</h4>
<p>Here, we use the function <code>simulateMetaTranscriptome()</code>. This function takes the reads assigned to each genome and the list of genes within this genome (extracted from the FASTA file), and then distributes the reads among the different genes of that genome. This function does this for the 5 genomes in our list, and for the 10 samples, and aggregates the output, resulting in a single gene count matrix, in which each row represents a single gene from any of this 5 bacteria and each column represents one of the 10 samples. This function also simulates differential expression. In this case, after generating the gene count matrix, it will assign any of the <code>fold_changes</code> randomly (according to the specified <code>fold_probs</code>) to each of the genes, and it will multiply the corresponding case columns for a gene by its assigned fold change. Therefore more arguments need to be specified in this case.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">metaT_experiment=<span class="kw">simulateMetaTranscriptome</span>(<span class="dt">genomeFileDir =</span> path_to_fastas, <span class="dt">genomeReadMatrix =</span> genomeComp, 
                                           <span class="dt">DE =</span> T, <span class="dt">foldChanges =</span> fold_changes, <span class="dt">foldProbs =</span> fold_probs, 
                                           <span class="dt">nSamples =</span> nCases,  <span class="dt">nControls =</span> nControls)
<span class="co">#&gt; Generating data from baseline model.</span>
<span class="co">#&gt; Generating data from baseline model.</span>
<span class="co">#&gt; Generating data from baseline model.</span>
<span class="co">#&gt; Generating data from baseline model.</span>
<span class="co">#&gt; Generating data from baseline model.</span></code></pre></div>
</div>
<div id="step-3-generate-the-fasta-files.-3" class="section level4">
<h4>Step 3: Generate the FASTA files.</h4>
<p>Here, we use the function <code>simulateFastaReads()</code>. This function takes the output from the previous one, as well as the gene FASTAS from each genome. It writes the simulated reads fasta files to the output directory (2 fasta files per sample in paired-end mode or 1 in single-end mode). It also writes the gene count matrix in the same directory, and a list of differentially expressed genes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Step 4: Generate the fasta files (function simulateFastaReads)
<span class="kw">simulateFastaReads</span>(<span class="dt">genomeFileDir =</span> path_to_fastas, <span class="dt">outdir =</span> output_simulation, 
                   <span class="dt">simulatedDataSet =</span> metaT_experiment)</code></pre></div>
<p>Similar code can be easily written to generate simulated datasets with different microbial composition AND differential expression, by combining examples 2/3 with example 4.</p>
</div>
</div>
</div>
<div id="session-info" class="section level1">
<h1>Session info:</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()
<span class="co">#&gt; R version 3.4.3 (2017-11-30)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin15.6.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS High Sierra 10.13.1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRblas.0.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] metaester_0.0.0.9000</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] Rcpp_0.12.17        knitr_1.20          XVector_0.18.0     </span>
<span class="co">#&gt;  [4] magrittr_1.5        splines_3.4.3       fitdistrplus_1.0-9 </span>
<span class="co">#&gt;  [7] MASS_7.3-50         BiocGenerics_0.24.0 zlibbioc_1.24.0    </span>
<span class="co">#&gt; [10] IRanges_2.12.0      lattice_0.20-35     logspline_2.1.11   </span>
<span class="co">#&gt; [13] stringr_1.3.1       tools_3.4.3         grid_3.4.3         </span>
<span class="co">#&gt; [16] parallel_3.4.3      polyester_1.9.7     seqinr_3.4-5       </span>
<span class="co">#&gt; [19] htmltools_0.3.6     survival_2.42-3     ade4_1.7-11        </span>
<span class="co">#&gt; [22] yaml_2.1.19         rprojroot_1.3-2     digest_0.6.15      </span>
<span class="co">#&gt; [25] Matrix_1.2-14       S4Vectors_0.16.0    evaluate_0.10.1    </span>
<span class="co">#&gt; [28] rmarkdown_1.10      limma_3.34.9        stringi_1.2.3      </span>
<span class="co">#&gt; [31] pasilla_1.6.0       compiler_3.4.3      backports_1.1.2    </span>
<span class="co">#&gt; [34] Biostrings_2.46.0   stats4_3.4.3</span></code></pre></div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
